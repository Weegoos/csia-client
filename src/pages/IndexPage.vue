<template>
  <div>
    <section align="right">
      <q-btn
        class="bg-grey q-mr-sm"
        round
        color="black"
        icon="mdi-bell-outline"
        @click="pushToNofications"
      />
      <q-btn
        style="background-color: #30322e"
        round
        icon="mdi-book-open-outline"
        @click="pushToGuide"
        class="q-mr-sm"
      />
      <q-btn
        style="background-color: #30322e"
        round
        icon="mdi-account-check"
        @click="pushToAdminPage"
      />
    </section>
    <p class="text-h5 text-bold" data-testid="indexPageMainText">
      {{ t("indexPage.mainText") }}
    </p>
    <span class="text-body1" data-testid="indexPageSubText">
      {{ t("indexPage.captionText") }}
    </span>
    <section
      class="q-pa-sm q-my-md"
      style="background-color: #30322e; border-radius: 8px"
      v-if="allMyPlants.content && allMyPlants.content.length === 0"
      data-testid="indexPageNoPlantsText"
    >
      <p class="text-body2 text-bold">{{ t("indexPage.addPlant") }}</p>
      <p>
        {{ t("indexPage.buttonCaptionText") }}
      </p>
    </section>
    <section v-else data-testid="indexPagePlantsText">
      <div>
        <q-list bordered>
          <q-item
            clickable
            v-ripple
            v-for="(plant, index) in allMyPlants.content"
            :key="index"
            @click="openDetailedMyPlant(plant)"
          >
            <q-item-section avatar class="text-white">
              <q-icon color="green-4" name="mdi-leaf" />
            </q-item-section>
            <q-item-section>
              <q-item-label class="text-body1 text-bold">{{
                plant.custom_name
              }}</q-item-label>
              <q-item-label class="text-white" caption lines="2">{{
                plant.substrate
              }}</q-item-label>
              <q-item-label class="text-white" caption lines="2">
                {{ t("indexPage.collectTime") }}:
                {{ plant.collectTime }}
              </q-item-label>
            </q-item-section>
            <q-item-section side center>
              <q-icon name="mdi-arrow-right" color="white" />
            </q-item-section>
          </q-item>
        </q-list>
      </div>
      <q-pagination
        class="justify-center"
        v-model="current"
        color="green-4"
        :min="1"
        :max="maxPage"
        @update:model-value="pagination"
        :boundary-numbers="true"
      />
    </section>
    <q-btn
      color="green-4 q-mt-md"
      rounded
      style="width: 100%"
      :label="t('indexPage.addPlantBtnLabel')"
      no-caps
      @click="addPlant"
    />
    <ViewDetailedInformationAboutPlant
      :openWindow="openWindow"
      :allInfoAboutPlant="allInfoAboutPlant"
      @closeWindow="closeWindow"
    />
  </div>
</template>

<script>
import { useQuasar } from "quasar";
import { getMethod } from "src/composables/apiMethod/get";
import ViewDetailedInformationAboutPlant from "../components/myPlant/ViewDetailedInformationAboutPlant.vue";
import { checkAccessToken } from "src/composables/javascript-function/token";
import { getCurrentInstance, onMounted, ref, watch } from "vue";
import { useRouter } from "vue-router";
import { useI18n } from "vue-i18n";
export default {
  components: {
    ViewDetailedInformationAboutPlant,
  },
  setup() {
    // global variables
    const router = useRouter();
    const { proxy } = getCurrentInstance();
    const serverURL = proxy.$serverURL;
    const maxItemsPerPage = proxy.$maxItemsPerPage;
    const $q = useQuasar();
    const { t } = useI18n();

    const pushToNofications = () => {
      router.push("/notifications");
    };

    const addPlant = () => {
      router.push("/create-plant");
    };

    const pushToGuide = () => {
      router.push("/guide");
    };

    const allInfoAboutPlant = ref([]);
    const openWindow = ref(false);
    const openDetailedMyPlant = (info) => {
      allInfoAboutPlant.value = info;
      openWindow.value = true;
    };

    const closeWindow = () => {
      openWindow.value = false;
    };

    const allMyPlants = ref([]);
    const getAllMyPlants = async (page) => {
      await getMethod(
        serverURL,
        `user/myPlants?page=${page}&size=${maxItemsPerPage}`,
        allMyPlants,
        $q,
        "Error: "
      );

      console.log(allMyPlants.value);
    };

    // pagination
    const current = ref(1);
    const pagination = (page) => {
      console.log("Текущая страница:", page);
      current.value = page;
      console.log(current.value);

      getAllMyPlants(current.value - 1);
    };

    const maxPage = ref("");
    watch(
      () => allMyPlants.value,
      (newVal) => {
        if (newVal && newVal.page.size) {
          maxPage.value = Math.ceil(
            newVal.page.totalElements / newVal.page.size
          );
        } else {
          maxPage.value = 1;
        }
      }
    );

    onMounted(() => {
      getAllMyPlants(current.value - 1);
      checkAccessToken();
    });

    const pushToAdminPage = () => {
      router.push("/admin/users");
    };

    return {
      pushToNofications,
      addPlant,
      pushToGuide,
      openDetailedMyPlant,
      allMyPlants,
      openWindow,
      allInfoAboutPlant,
      closeWindow,
      current,
      pagination,
      maxPage,
      pushToAdminPage,
      t,
    };
  },
};
</script>

<style></style>
