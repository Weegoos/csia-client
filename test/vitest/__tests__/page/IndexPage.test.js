import { installQuasarPlugin } from "@quasar/quasar-app-extension-testing-unit-vitest";
import { mount } from "@vue/test-utils";
import { Quasar } from "quasar";
import IndexPage from "src/pages/IndexPage.vue";
import { describe, expect, it, vi } from "vitest";
import { getMethod } from "src/composables/apiMethod/get";

installQuasarPlugin();

let testData = [
  {
    id: "67e84d8a26054e6d4674640e",
    name: "Rukkola",
    type: "Herbaceous plant",
    difficulty: "Easy",
    characteristic: {
      water: {
        when: "Water when the topsoil is dry",
        tip: "Prefers regular but not heavy watering",
        difficulty: "Easy",
      },
      light: {
        preferredLight: "Sunny location or partial shade",
        suitableWith: "Suitable for south or east-facing windows",
      },
      toxicity: {
        toxicTo: "Non-toxic",
        power: "Safe for pets",
      },
      humidity: {
        scale: "Moderate humidity",
        ratio: "40–60%",
      },
      common_problems: {
        name: "Root rot",
        detail: "Occurs if soil is overwatered",
      },
      common_pests: {
        name: "Aphids",
        detail: "Often found on young leaves",
      },
      liquid_fertilizing: "Fertilize every 2 weeks with organic fertilizer",
      temperatures: {
        sentence: "Grows best in temperatures between 10°C and 25°C",
        start: 10,
        end: 25,
        indoorMonths: "October – March",
        outDoorMonths: "April – September",
      },
      clean: "Remove dry and yellowing leaves regularly",
      leaves: {
        color: "Green",
        type: "Pinnate, narrow",
      },
      dimensions: {
        beginHeight: 10,
        endHeight: 30,
        beginSpread: 5,
        endSpread: 20,
      },
    },
    images: ["67e85dbd4539f1454e250655", "67e85e014539f1454e250657"],
  },
  {
    id: "67e84fccf009297bde2d8e38",
    name: "Watercress",
    type: "Herbaceous plant",
    difficulty: "Moderate",
    characteristic: {
      water: {
        when: "Keep soil consistently moist, but not waterlogged",
        tip: "Requires frequent light watering, especially during germination",
        difficulty: "Moderate",
      },
      light: {
        preferredLight: "Bright, indirect sunlight or partial shade",
        suitableWith: "Suitable for east or north-facing windows",
      },
      toxicity: {
        toxicTo: "Non-toxic",
        power: "Safe for pets and children",
      },
      humidity: {
        scale: "Moderate to high humidity",
        ratio: "50–70%",
      },
      common_problems: {
        name: "Wilting",
        detail:
          "Can occur if the soil dries out too quickly or receives too much direct sun",
      },
      common_pests: {
        name: "Flea beetles",
        detail: "May create small holes in leaves",
      },
      liquid_fertilizing:
        "Fertilize every 3 weeks with diluted organic fertilizer",
      temperatures: {
        sentence: "Thrives in cool temperatures between 10°C and 20°C",
        start: 10,
        end: 20,
        indoorMonths: "October – March",
        outDoorMonths: "April – September",
      },
      clean:
        "Trim overgrown stems and remove yellow leaves to encourage fresh growth",
      leaves: {
        color: "Green",
        type: "Small, oval to slightly serrated",
      },
      dimensions: {
        beginHeight: 5,
        endHeight: 15,
        beginSpread: 3,
        endSpread: 10,
      },
    },
    images: [],
  },
];

vi.mock("src/composables/apiMethod/get", () => ({
  getMethod: vi.fn(() => Promise.resolve({ data: testData })),
}));

describe("tests for IndexPage", () => {
  const wrapper = mount(IndexPage, {
    global: {
      plugins: [Quasar],
    },
  });
  it("should find indexPageMainText data-testid", () => {
    const indexPageMainText = wrapper.find('[data-testid="indexPageMainText"]');
    expect(indexPageMainText.exists()).toBe(true);
    expect(indexPageMainText.text()).toContain("Good");
    expect(indexPageMainText).toMatchSnapshot();
  });
});
