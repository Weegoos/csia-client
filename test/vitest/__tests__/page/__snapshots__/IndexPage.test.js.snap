// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tests for IndexPage > should find indexPageMainText data-testid 1`] = `
DOMWrapper {
  "isDisabled": [Function],
  "wrapperElement": <p
    class="text-h5 text-bold"
    data-testid="indexPageMainText"
  >
    Good day
  </p>,
}
`;

exports[`tests for IndexPage > should find indexPageSubText data-testid 1`] = `
DOMWrapper {
  "isDisabled": [Function],
  "wrapperElement": <span
    class="text-body1"
    data-testid="indexPageSubText"
  >
     Your plants are so glad you are here.
  </span>,
}
`;
